src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,109 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] objects;
+    private int size;
+
+    public ArrayList() {
+        objects = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= objects.length) {
+            grow();
+        }
+        objects[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (size >= objects.length) {
+            this.grow();
+        }
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("index: " + index
+                    + " out of bounds for size: " + size);
+        }
+        Object[] newObjectsArray = new Object[objects.length + 1];
+        System.arraycopy(objects, 0, newObjectsArray, 0, index);
+        newObjectsArray[index] = value;
+        System.arraycopy(objects, index, newObjectsArray, index + 1, objects.length - index);
+        objects = newObjectsArray;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) objects[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        objects[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Object[] newObjectsArray = new Object[objects.length];
+        System.arraycopy(objects, 0, newObjectsArray, 0, index);
+        System.arraycopy(objects, index + 1, newObjectsArray, index, objects.length - (index + 1));
+        T removedObject = (T) objects[index];
+        objects = newObjectsArray;
+        size--;
+        return removedObject;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if (element == objects[i]
+                    || (objects[i] != null && objects[i].equals(element))) {
+                index = i;
```

Comment:
```suggestion
                index = i;
                break;
```
if you find the index you need, you can easily break, to stop the for-loop